### PR TITLE
Docker image for AWS provisioning

### PR DESCRIPTION
**Pull request by stepanstipl**
I've created a docker image which should simplify bringing up AWS infrastructure. How to run this image is documented in README.

```
docker run -it --rm -e USER -e AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY -e AWS_REGION \
            -v $(pwd)/test.pub:/root/.ssh/id_rsa.gpii-ci.pub \
            -v $(pwd)/test:/root/.ssh/id_rsa.gpii-ci \
            gpii/gpii-infra-aws rake
```

The image `gpii/gpii-infra-aws` has been pushed manually, once this is merged we can setup automated build.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
+FROM ruby:2.4.4-alpine3.7
```
Comment (mrtyler) [inline Dockerfile]: Can this move to aws/Dockerfile, or will that cause trouble with the Docker Hub integration you were trying to set up?

Comment (stepanstipl) [reply Dockerfile]: Unfortunately seems like Docker Hub can't handle this. (if running manually the build would have to be called smth like `docker build -t xyz ../`, basically the context have to be the root directory, as we need the whole repo inside)

**Comment by stepanstipl** [comment]
You're right about the checksums, I'm happy to add those if we plan to use it for more than devs tool.

My intention was primarily for dev convenience at this stage. If we should take it further and use it for CI, given we're moving away from AWS infra, I'm not sure if it's worth it.

Probably would have to solve some bits - like the credentials issue and also you mentioned possible docker-in-docker problem with some tests?

File: aws/README.md [hunk 2]
```diff
@@ -73,6 +73,27 @@ I generally download and install these tools by hand (verifying checksums when a
...
+- `-v $(pwd)/test:/root/.ssh/id_rsa.gpii-ci`  mount the ssh private key
+- `-v $(pwd)/test.pub:/root/.ssh/id_rsa.gpii-ci.pub`  mount the ssh public key
+- (*optional*) if you want to run current code, instead the one in docker image,
+  add mount `-v "$(pwd)/..:/gpii-infra"`
+
+*__CAVEAT__: There's known issue with DB credentials, which are generated by Rake
+task and not stored in TF state, so if your provisioning fails during the run, these
+might be recreated and cause an error. Partial solution is to run with  `-v
+"$(pwd)/..:/gpii-infra"` or start interactive terminal and run the rake task
+from there, so that the container won't get deleted between runs.
```
Comment (amatas) [inline aws/README.md]: Also, if you set the secrets as environment variables it should work. It's what we do at the CI.

File: Dockerfile (hunk 1 above)
Comment (mrtyler) [reply Dockerfile]: Ok.

**Comment by mrtyler** [comment]
If the current target is as a developer convenience, this LGTM (though Alfredo raises some good points about handling secrets more seamlessly).

I think it's fine to let people try this out and see how it works, then modify CI to use it if that experiment is successful (and if we think it's worth the investment on the AWS side).

**Comment by stepanstipl** [comment]
@amatas thanks, re. docker-compose - I'll add the compose file, definitely makes sense for dev use case. For any CI/CD/production use, I would prefer to leave compose out of the game, as it adds on complexity & in my experience compose has a lot of issues.

re. files - can you point me to some specific example? I know just about the secrets file so far. I think ideally the same applies - we should try to avoid having state split across TF remote one and local files.

**Comment by stepanstipl** [comment]
Added `docker-compose` so now you can run
```
# to create
docker-compose run gpii-infra-aws "rake"
# to destroy
docker-compose run gpii-infra-aws "rake destroy && rake clobber
```

Looking into the secrets issue.

**Comment by stepanstipl** [comment]
@amatas re. secrets, I guess it's the following ones, right? https://github.com/gpii-ops/gpii-infra/blob/e42661e8aa2885e98ccd5608ec3e5c6525ffc533/aws/modules/deploy/Rakefile#L48-L55

I guess most of those are not even secrets (hosts, ports..), so those can be easily added as env variables (although I would ask why not set some sane defaults/hardcode at a first place).

I would be worried about passwords though - one option is to put these `into docker-compose.yaml` (or .env) file, but that would result in having same password on every deployment - I'm not sure I like this. Any ideas?

**Comment by stepanstipl** [comment]
Thanks, Sergey. I would like to avoid having external dependencies on Ruby, Rake & Bundler, I think that takes half of the usefulness of packing stuff in Docker away (the readme asks for very specific versions of Ruby and other dependencies and it can turn out to be non-trivial on many systems, unless you start using smth. like rvm). At least in my case main reason to create the docker image was to avoid having to install these.

**Comment by natarajaya** [comment]
Well, I think Ruby and Rake are pretty light dependencies. For most of the systems installing them would look something like `you_package_manager install ruby >= 2.0.0 && gem install bundler && bundle`. And it is just 2 out of [10 dependencies](https://github.com/gpii-ops/gpii-infra/tree/master/aws#install-packages), so having the rest of them in Docker image still makes a lot of sense and will make our life easier.

**Comment by stepanstipl** [comment]
Thanks, Sergey. Not so simple on Mac, as it does not have package manager and comes with ruby 2.3.3 by default. We ask specifically for `2.4.3` in the readme and also a specific version of bundler and rake too `12.3.0`. Also ruby calls a couple of tools directly (I think it's least jq, curl & ssh).

Mainly though, this would prevent the use case when you actually don't need to check out the git code at all, as everything is already in the docker image.

**Comment by natarajaya** [comment]
Mac got its own "de facto standard" package manager - Homebrew :)

Not sure if its any useful from developers perspective to have code inside docker image, since we usually make changes into it before running. From CI perspective, it also seems more convenient to update image only when there is a change in tooling, and not every time there is a change in code.

@mrtyler I don't see a reason why we need specific versions of ruby and bundler. As I get from `rake` [gem requirements](https://rubygems.org/gems/rake), anything  >= `2.0.0` should work. Tools that is called directly from ruby code is different question, but I think beside "standard" system utilities, there is only one more dependency on `jq`.

**Comment by mrtyler** [comment]
Why we have such a specific version requirement:
* Fewer supported versions means fewer compatibility problems
* We have painful experience with version incompatibilities between even point releases of tools like terraform and kops

That said, I agree that ruby, rake, and bundler are low risk. If relaxing version constraints on those tools makes life easier, that's fine with me.

**Comment by stepanstipl** [comment]
Relaxing version constraints on Rake & co. wouldn't really help, I'm just working on another ticket where the problem is caused by differences between OS X version of curl and openssl vs. linux ones.

**Comment by stepanstipl** [comment]
I think the work is useful as it is already, I'm successfully using the image to create and tear down clusters.

However, I don't want to invest too much more work into this, as it was intended for dev convenience and I think all the approaches are non-trivial.

- Hard-coding values as env-variables is imho undesirable and would lead into having clusters with the same secrets
- Leaving Rake out of the container - this is is against the idea of this whole solution, and as we can see for example in #82, it's not just Ruby, rake & bundler, but also things like curl or jq
  - This would also require modifying current rake code on quite a few places and I would anticipate more issues to emerge

Let me know if you think it's worth merging in the current state, otherwise, I suggest to close this PR.

**Comment by mrtyler** [comment]
If it is usable in its current state (even with caveats, as long as the caveats are documented), I'm fine with merging it and seeing if it becomes popular. If you plan to continue using it, let's have a copy in revision control instead of only on your laptop ;).

I'm also fine with closing this and minimizing clutter in the repo, if you think that's better.

**Comment by natarajaya** [comment]
@stepanstipl We should either address secrets issue or abandon this, imo. 
With current state of things, developer will always have to rely on host machine ENV variables to populate secrets, which is not very convenient and may lead to various issues (i.e accidentally messing up clusters by running deploy from another terminal tab with ENV vars unset, etc.)

**Comment by stepanstipl** [comment]
@natarajaya This issue does not happen if you mount the local checked-out directory using `-v
"$(pwd)/..:/gpii-infra"` or use the bundled docker-compose. Do you think it's still not good enough? 

(you mentioned previoulsy that *"Not sure if its any useful from developers perspective to have code inside docker image"*, so I assume you don't consider use case of not doing so very common?)

**Comment by stepanstipl** [comment]
I haven't received any feedback on this and looks like we're gonna drop AWS soon, so I'm closing this for now.

**Pull request closed without merging**